iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -0,0 +1,317 @@
+// Copyright 2020 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringRef.h"
+#include "mlir/Conversion/StandardToLLVM/ConvertStandardToLLVM.h"
+#include "mlir/Conversion/VectorToLLVM/ConvertVectorToLLVM.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+// Returns true if `contractionOp` is of the form
+//   matrix * transposed_matrix.
+// That is, if there are 2 parallel iterators, say M and N, 1 additive reduction
+// iterator, say K, and the indexing maps are {{M, K}, {N, K}, {M, N}}.
+static bool isMatrixTimesMatrixTransposed(vector::ContractionOp contractionOp) {
+  // Check that the reduction is additive.
+  if (contractionOp.kind() != vector::CombiningKind::ADD) {
+    return false;
+  }
+  // Check that there are 2 parallel and 1 reduction iterators.
+  auto iteratorTypes = contractionOp.iterator_types().getValue();
+  if (iteratorTypes.size() != 3) {
+    return false;
+  }
+  SmallVector<int, 3> parallel_iterators;
+  SmallVector<int, 3> reduction_iterators;
+  for (int i = 0; i < 3; i++) {
+    if (isParallelIterator(iteratorTypes[i])) {
+      parallel_iterators.push_back(i);
+    } else if (isReductionIterator(iteratorTypes[i])) {
+      reduction_iterators.push_back(i);
+    } else {
+      return false;
+    }
+  }
+  if (parallel_iterators.size() != 2 || reduction_iterators.size() != 1) {
+    return false;
+  }
+  // Give the found iterators some idiomatic names.
+  const int M = parallel_iterators[0];
+  const int N = parallel_iterators[1];
+  const int K = reduction_iterators[0];
+  // Check that there are 3 indexing maps.
+  auto indexingMaps = contractionOp.indexing_maps().getValue();
+  if (indexingMaps.size() != 3) {
+    return false;
+  }
+  // Check that the indexing maps have the expected form.
```

Comment:
You can just build the expected maps and check that they are the same map. `AffineMap`s are uniqued within a context. So equality can be checked on `AffineMap` objects.

Reply:
The problem is that a Matrix-times-Matrix-transpose can be represented in more than one way as a vector.contract, so there isn't a single tuple of affinemaps that I could compare to, unless we wanted to support only special cases. I can see how that would be acceptable as an implementation-effort tradeoff, but here I found it just as simple to write more pedestrian C-ish code handling a more general case.